### ISSUE TITLE
fpdart v1.0.0

### ISSUE THREAD

**Issue by SandroMaglione** [SandroMaglione/fpdart#108]
[Dart 3 is out](https://medium.com/dartlang/announcing-dart-3-53f065a10635), and with it new features like pattern matching, records, and class modifiers.

These features are making dart safer and more "functional" 🎯

It's time to upgrade `fpdart` to support Dart 3 features. Refactoring `fpdart` with patterns, sealed classes, and records will bring many more advancements, but also many breaking changes.

Therefore, `fpdart` v1.0.0 will raise the min sdk to Dart 3, deprecate some APIs (e.g. `Tuple2`), and add some new features.

This issue is used to discuss any possible new features to release as part of `fpdart` v1.0.0.

## Feel free to drop below any suggestion, feedback, or request for things that you would want to have (or remove) as part of `fpdart` v1.0.0 👇

**Comment by cranst0n** [comment]
I've recently published an set of FP related packages using Dart 3. While I use them for work and hacking on, I don't have any interest in creating yet another FP library since there are already plenty of good ones around, fpdart among them. I think there are some things in what I've built that could be useful additions to fpdart.

You can find the repo [here](https://github.com/cranst0n/ribs)

Among things it includes that may be worth considering for fpdart:
* [IO](https://github.com/cranst0n/ribs/blob/main/packages/core/lib/src/effect/io.dart) monad closely derived from the [cats-effect IO](https://github.com/typelevel/cats-effect/blob/series/3.x/core/shared/src/main/scala/cats/effect/IO.scala)
* [Retry cababilities](https://github.com/cranst0n/ribs/blob/main/packages/core/lib/src/effect/io_retry.dart) for `IO` closely derived from [cats-retry](https://github.com/cb372/cats-retry)
* [IList](https://github.com/cranst0n/ribs/blob/main/packages/core/lib/src/ilist.dart) and related [NonEmptyIList](https://github.com/cranst0n/ribs/blob/main/packages/core/lib/src/non_empty_ilist.dart). These ended up being wrappers around the excellent [FIC](https://pub.dev/packages/fast_immutable_collections) IList for performance reasons, so their addition is probably more debatable
* [Validated](https://github.com/cranst0n/ribs/blob/main/packages/core/lib/src/validated.dart)
* Various [syntax extensions](https://github.com/cranst0n/ribs/tree/main/packages/core/lib/src/syntax) leaning on Dart 3 tuples
* [JSON library](https://github.com/cranst0n/ribs/blob/main/packages/json) closely derived from [circe](https://github.com/circe/circe)
* [Binary codec library](https://github.com/cranst0n/ribs/tree/main/packages/binary) closely derived from [scodec](https://github.com/scodec/scodec)
* [Lens library](https://github.com/cranst0n/ribs/tree/main/packages/optics) closely derived from [monocle](https://github.com/optics-dev/Monocle)
* [Units library](https://github.com/cranst0n/ribs/tree/main/packages/units) closely derived from [squants](https://github.com/typelevel/squants)

The main limitation of what I've created is the lack of documentation, and in some cases the lack of thorough tests, both of which, fpdart excels at.

I also [made an attempt](https://github.com/cranst0n/ribs/tree/main/packages/rill) to port [fs2](https://github.com/typelevel/fs2) to dart but ran into limitations in the dart type system (or maybe my own limited knowledge of it).

Maybe some of this stuff would be a good fit in `fpdart` but at the very least, I hope it can provide some ideas on how to move to v1.0.0.

**Comment by RandalSchwartz** [comment]
Just off the top of my head, but I was thinking that an easy way to go from the existing Tuple monad back and forth to a Record (T1, T2) would be great.  We can never turn the records into monads, so there's always going to need to be a way to lift from records into proper monad space.

**Comment by SandroMaglione** [comment]
@RandalSchwartz I think that in practice using Records would lead to a better developer experience and easier maintainability than `Tuple` in dart 3.

What practical use do you see for having `Tuple` as monad? What do you think would be a good reason to have and use `Tuple` instead of Records in Dart 3?

**Comment by RandalSchwartz** [comment]
Another thought... I think your next release should be still under 1.0, because there will likely need to be a few breaking changes while you are absorbing the proper best practices about the new features.  I'd hate there to be version 2.0 in six months, but I'm willing to live with 0.9 and 0.9.1 etc.

**Comment by RandalSchwartz** [comment]
> Update: Removing `Tuple2` aligns better with native dart code, as well as making the code less verbose and reducing lines of code in `fpdart` ([PR](https://github.com/SandroMaglione/fpdart/pull/112))

Yes, I completely agree, looking at the work you did in the past few days.  Much cleaner to not even need the artificial class.  It just means I need to do some work for one of my clients to advance fpdart for them. :)

**Comment by FaFre** [comment]
I would really like to have the option to only import the core functionality of fpdart like `Either`, `Option` etc.

I only use a small subset of the fpdart library and sometimes I accidentally use unintentionally the exposed fpdart extensions for `List`, `Iterable` etc. which I don't intent to. E.g. fpdart's `all` instead of the dart core's `every`. Confusion sometimes happens when you code in multiple languages.

I know that I would have the possibility to work with `hide` and `show`, but a more comfortable solution would be maybe having a specific import like `import 'package:fpdart/core.dart';` for that.

**Comment by SandroMaglione** [comment]
@FaFre how would you suggest to split imports? What do you think `core.dart` should contain?

**Comment by FaFre** [comment]
Everything, but global extensions for types that are not having their origin in fpdart themselves.

**Comment by RandalSchwartz** [comment]
you can have fpdart/fpdart.dart continue to be the whole thing, but something like fpdart/core.dart be only the non-extension types.

**Comment by RandalSchwartz** [comment]
> Everything, but global extensions for types that are not having their origin in fpdart themselves.

I like this.  And yes, I'd call "core" everything but the extensions.  I presume that none of the extensions are used by any of the core.

**Comment by SandroMaglione** [comment]
The first beta release ([v1.0.0-beta.1](https://pub.dev/packages/fpdart/versions/1.0.0-beta.1)) has been released.

Feel free to provide any feedback and report any issue here 👇🙏🏼

**Comment by Musta-Pollo** [comment]
Thank you for the incredible update ❤️! 

I wanted to bring to your attention that the links to the examples in the documentation are not functional. An example can be found here:

![image](https://github.com/SandroMaglione/fpdart/assets/68862752/5ed930a7-debe-4ac4-bd30-b3a501067ee5)

**Comment by SandroMaglione** [comment]
@Musta-Pollo thanks for reporting, this should have been fixed in #122

**Comment by TC3000** [comment]
Could we please have a better way of handling streams in V1? There was a suggested approach here https://github.com/SandroMaglione/fpdart/issues/27#issuecomment-1338654212, curious to hear your thoughts.

**Comment by AdamHavlicek** [comment]
hello, why id was renamed to identity? it adds unnecessary breaking change to api.

**Comment by Musta-Pollo** [comment]
> hello, why id was renamed to identity? it adds unnecessary breaking change to api.

Personally, I am glad about this change because I had a lot of name conflicts when importing the fpdart. I would need to use:

```import '...fpdart/fpdart.dart' hide id;```
And that was not ideal.

**Comment by SandroMaglione** [comment]
@AdamHavlicek actually is not that `id` was renamed to `identity`, but that fpdart previously had both `id` and `identity` which did the same thing. As @Musta-Pollo says, in order to avoid conflicts and confusion with the same function with 2 different names, `id` has been remove in favour of `identity`.

**Comment by SandroMaglione** [comment]
Hi everyone,

I am aiming for releasing `fpdart` v1.0.0 next week 🚀

Comment below for any last suggestion or issue 👇

**Comment by FaFre** [comment]
Will you address the suggested change mentioned in https://github.com/SandroMaglione/fpdart/issues/108#issuecomment-1556138403 ?

**Comment by SandroMaglione** [comment]
@FaFre I would prefer not to add multiple imports to avoid adding complexity when setting up and using the library. Many people that use the library for the first time I think would benefit to have everything at their disposal, especially since you can learn about a new functionality by using auto-complete.

Furthermore, in v1 the extension methods have been refactored to improve performance.

**Comment by RandalSchwartz** [comment]
:shipit: !!!!
